File: 0198-house-robber/0198-house-robber.py
class Solution:
    def rob(self, nums: List[int]) -> int:
        n = len(nums)
        
        if n == 0:
            return 0
        if n == 1:
            return nums[0]

        prev2 = 0
        prev1 = nums[0]

        for i in range(2, n + 1):
            steal = nums[i-1] + prev2
            skip = prev1
            current = max(steal, skip)
            prev2 = prev1
            prev1 = current

        return current

File: 0198-house-robber/0198-house-robber.java
class Solution {
    public int rob(int[] nums) {
        int n = nums.length;
        int[] money = new int[n + 1];
        money[0] = 0;
        money[1] = nums[0];
        
        for (int i = 2; i <= n; i++) {
            int steal = nums[i-1] + money[i-2];
            int skip = money[i-1];
            money[i] = Math.max(steal, skip);
        }
        
        return money[n];
    }
}